File: tests/catch2_tests/test_qdant_read.cpp
// SPDX-License-Identifier: Apache-2.0
//
// quadriga-lib c++/MEX Utility library for radio channel modelling and simulations
// Copyright (C) 2022-2023 Stephan Jaeckel (https://sjc-wireless.com)
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ------------------------------------------------------------------------

#include <catch2/catch_test_macros.hpp>

#include "quadriga_lib.hpp"
#include "quadriga_tools.hpp"

#include <iostream>
#include <fstream>
#include <string>

TEST_CASE("Arrayant reading from QDANT file - Minimal test")
{
    std::ofstream f;
    f.open("test.qdant");
    f << "<qdant><arrayant>" << std::endl;
    f << "<name>bla</name>" << std::endl;
    f << "<ElevationGrid>-90 -45 0 45 90</ElevationGrid>" << std::endl;
    f << "<AzimuthGrid>-180 0 90</AzimuthGrid>" << std::endl;
    f << "<EthetaMag>1 2 3 4 5 6 7 8 9 10 11 12 13 14 15</EthetaMag>" << std::endl;
    f << "</arrayant></qdant>" << std::endl;
    f.close();

    float pi = arma::datum::pi;
    auto x = quadriga_lib::qdant_read<float>("test.qdant");

    arma::fmat A = 20 * log10(x.e_theta_re.slice(0));
    arma::fmat B = arma::reshape(arma::linspace<arma::fvec>(1, 15, 15), 3, 5).t();
    CHECK(arma::approx_equal(A, B, "absdiff", 1e-6));

    B.zeros();
    CHECK(arma::approx_equal(x.e_theta_im.slice(0), B, "absdiff", 1e-6));
    CHECK(arma::approx_equal(x.e_phi_re.slice(0), B, "absdiff", 1e-6));
    CHECK(arma::approx_equal(x.e_phi_im.slice(0), B, "absdiff", 1e-6));

    B = {-pi, 0.0f, pi / 2.0f};
    CHECK(arma::approx_equal(x.azimuth_grid, B.t(), "absdiff", 1e-6));

    B = {-pi / 2.0f, -pi / 4.0f, 0.0f, pi / 4.0f, pi / 2.0f};
    CHECK(arma::approx_equal(x.elevation_grid, B.t(), "absdiff", 1e-6));

    B = {0.0f, 0.0f, 0.0f};
    CHECK(arma::approx_equal(x.element_pos, B.t(), "absdiff", 1e-6));

    B = {1.0f};
    CHECK(arma::approx_equal(x.coupling_re, B, "absdiff", 1e-6));

    B = {0.0f};
    CHECK(arma::approx_equal(x.coupling_im, B, "absdiff", 1e-6));
    CHECK(std::abs(x.center_frequency - 299792447.0f) < 0.1f);

    std::remove("test.qdant");
}

TEST_CASE("Arrayant reading from QDANT file - Multi-Element Array")
{
    std::ofstream f;
    f.open("test2.qdant");
    f << "<qdant><arrayant>" << std::endl;
    f << "<name>bla</name>" << std::endl;
    f << "<NoElements>2</NoElements >" << std::endl;
    f << "<ElevationGrid>-90 -45 0 45 90</ElevationGrid>" << std::endl;
    f << "<AzimuthGrid>-180 0 90</AzimuthGrid>" << std::endl;
    f << "<EthetaMag   el=\"1\">1 2 3 4 5 6 7 8 9 10 11 12 13 14 15</EthetaMag>" << std::endl;
    f << "<EphiMag     el=\"1\">2 3 4 5 6 7 8 9 10 11 12 13 14 15 16</EphiMag>" << std::endl;
    f << "<EthetaMag   el=\"2\">3 4 5 6 7 8 9 10 11 12 13 14 15 16 17</EthetaMag>" << std::endl;
    f << "<EphiMag     el=\"2\">4 5 6 7 8 9 10 11 12 13 14 15 16 17 18</EphiMag>" << std::endl;
    f << "<EthetaPhase el=\"1\">1 1 1 1 1 1 1 1 1 1 1 1 1 1 1</EthetaPhase>" << std::endl;
    f << "<EphiPhase   el=\"1\">2 2 2 2 2 2 2 2 2 2 2 2 2 2 2</EphiPhase>" << std::endl;
    f << "<EthetaPhase el=\"2\">3 3 3 3 3 3 3 3 3 3 3 3 3 3 3</EthetaPhase>" << std::endl;
    f << "<EphiPhase   el=\"2\">4 4 4 4 4 4 4 4 4 4 4 4 4 4 4</EphiPhase>" << std::endl;
    f << "<CouplingAbs>1 2 3 4 5 6</CouplingAbs>" << std::endl;
    f << "<CouplingPhase>5 10 15 20 25 30</CouplingPhase>" << std::endl;
    f << "</arrayant></qdant>" << std::endl;
    f.close();

    double pi = arma::datum::pi;
    auto x = quadriga_lib::qdant_read<double>("test2.qdant");

    arma::mat A = 20 * log10(sqrt(x.e_theta_re.slice(0) % x.e_theta_re.slice(0) + x.e_theta_im.slice(0) % x.e_theta_im.slice(0)));
    arma::mat B = arma::reshape(arma::linspace<arma::vec>(1, 15, 15), 3, 5).t();
    CHECK(arma::approx_equal(A, B, "absdiff", 1e-13));

    A = 20 * log10(sqrt(x.e_phi_re.slice(0) % x.e_phi_re.slice(0) + x.e_phi_im.slice(0) % x.e_phi_im.slice(0)));
    B = arma::reshape(arma::linspace<arma::vec>(2, 16, 15), 3, 5).t();
    CHECK(arma::approx_equal(A, B, "absdiff", 1e-13));

    A = 20 * log10(sqrt(x.e_theta_re.slice(1) % x.e_theta_re.slice(1) + x.e_theta_im.slice(1) % x.e_theta_im.slice(1)));
    B = arma::reshape(arma::linspace<arma::vec>(3, 17, 15), 3, 5).t();
    CHECK(arma::approx_equal(A, B, "absdiff", 1e-13));

    A = 20 * log10(sqrt(x.e_phi_re.slice(1) % x.e_phi_re.slice(1) + x.e_phi_im.slice(1) % x.e_phi_im.slice(1)));
    B = arma::reshape(arma::linspace<arma::vec>(4, 18, 15), 3, 5).t();
    CHECK(arma::approx_equal(A, B, "absdiff", 1e-13));

    A = arma::atan2(x.e_theta_im.slice(0), x.e_theta_re.slice(0));
    B.fill(pi / 180.0);
    CHECK(arma::approx_equal(A, B, "absdiff", 1e-13));

    A = arma::atan2(x.e_phi_im.slice(0), x.e_phi_re.slice(0));
    B.fill(2.0 * pi / 180.0);
    CHECK(arma::approx_equal(A, B, "absdiff", 1e-13));

    A = arma::atan2(x.e_theta_im.slice(1), x.e_theta_re.slice(1));
    B.fill(3.0 * pi / 180.0);
    CHECK(arma::approx_equal(A, B, "absdiff", 1e-13));

    A = arma::atan2(x.e_phi_im.slice(1), x.e_phi_re.slice(1));
    B.fill(4.0 * pi / 180.0);
    CHECK(arma::approx_equal(A, B, "absdiff", 1e-13));

    A = sqrt(x.coupling_re % x.coupling_re + x.coupling_im % x.coupling_im);
    B = arma::reshape(arma::linspace<arma::vec>(1, 6, 6), 2, 3);
    CHECK(arma::approx_equal(A, B, "absdiff", 1e-13));

    A = arma::atan2(x.coupling_im, x.coupling_re) * 180.0 / pi;
    B = arma::reshape(arma::linspace<arma::vec>(5, 30, 6), 2, 3);
    CHECK(arma::approx_equal(A, B, "absdiff", 1e-12));

    std::remove("test2.qdant");
}

TEST_CASE("Arrayant reading from QDANT file - More complex test")
{
    std::ofstream f;
    f.open("test.qdant");
    f << "<?xml version=\"1.0\" encoding=\"UTF-8\"?><qdant xmlns:xx=\"test\">" << std::endl;
    f << "<xx:layout>1,1 1,1 1,1</xx:layout>" << std::endl;
    f << "<xx:arrayant id=\"1\">" << std::endl;
    f << "<xx:AzimuthGrid>-90 -45 0 45 90</xx:AzimuthGrid>" << std::endl;
    f << "<xx:ElevationGrid>-90 0 90</xx:ElevationGrid>" << std::endl;
    f << "<xx:EphiMag>0  0  0  0  0  0  0  0  0  0  0  0  0  0  0</xx:EphiMag>" << std::endl;
    f << "<xx:EphiPhase>90  90  90  90  90  90  90  90  90  90  90  90  90  90  90</xx:EphiPhase>" << std::endl;
    f << "<xx:EthetaMag>3  3  3  3  3  3  3  3  3  3  3  3  3  3  3</xx:EthetaMag>" << std::endl;
    f << "<xx:EthetaPhase>-90 -90 -90 -90 -90 -90 -90 -90 -90 -90 -90 -90 -90 -90 -90</xx:EthetaPhase>" << std::endl;
    f << "<xx:ElementPosition>1,2,3</xx:ElementPosition>" << std::endl;
    f << "<xx:CouplingAbs>1</xx:CouplingAbs>" << std::endl;
    f << "<xx:CouplingPhase>45</xx:CouplingPhase>" << std::endl;
    f << "<xx:CenterFrequency>3e9</xx:CenterFrequency>" << std::endl;
    f << "</xx:arrayant></qdant>" << std::endl;
    f.close();

    double pi = arma::datum::pi;

    arma::Mat<unsigned> layout;
    auto x = quadriga_lib::qdant_read<double>("test.qdant", 1, &layout);

    arma::mat B = {-pi / 2.0, -pi / 4.0, 0.0, pi / 4.0, pi / 2.0};
    CHECK(arma::approx_equal(x.azimuth_grid, B.t(), "absdiff", 1e-13));

    B = {-pi / 2.0, 0.0, pi / 2.0};
    CHECK(arma::approx_equal(x.elevation_grid, B.t(), "absdiff", 1e-13));

    B.zeros(3, 5);
    CHECK(arma::approx_equal(x.e_theta_re.slice(0), B, "absdiff", 1e-13));
    CHECK(arma::approx_equal(x.e_phi_re.slice(0), B, "absdiff", 1e-13));

    B.fill(-std::sqrt(std::pow(10, 0.3)));
    CHECK(arma::approx_equal(x.e_theta_im.slice(0), B, "absdiff", 1e-13));

    B.ones();
    CHECK(arma::approx_equal(x.e_phi_im.slice(0), B, "absdiff", 1e-13));

    B = {1.0, 2.0, 3.0};
    CHECK(arma::approx_equal(x.element_pos, B.t(), "absdiff", 1e-13));

    B = {1.0 / std::sqrt(2.0)};
    CHECK(arma::approx_equal(x.coupling_re, B, "absdiff", 1e-13));
    CHECK(arma::approx_equal(x.coupling_im, B, "absdiff", 1e-13));

    CHECK(std::abs(x.center_frequency - 3.0e9) < 0.1f);
    CHECK(x.name == "unknown");
    CHECK(arma::all(arma::vectorise(layout) == 1));

    std::remove("test.qdant");
}

File: tests/mex_tests/test_qdant_read.m
function test_qdant_read

%% Minimal test
f = fopen( 'test.qdant','w' );
fprintf(f,'%s\n','<qdant><arrayant>');
fprintf(f,'%s\n','<name>bla</name>');
fprintf(f,'%s\n','<ElevationGrid>-90 -45 0 45 90</ElevationGrid>');
fprintf(f,'%s\n','<AzimuthGrid>-180 0 90</AzimuthGrid>');
fprintf(f,'%s\n',['<EthetaMag>',num2str(1:15),'</EthetaMag>']);
fprintf(f,'%s\n','</arrayant></qdant>');
fclose(f);

[e_theta_re, e_theta_im, e_phi_re, e_phi_im, azimuth_grid, elevation_grid, element_pos, ...
    coupling_re, coupling_im, center_frequency, name, layout] = quadriga_lib.arrayant_qdant_read('test.qdant');

assertTrue( isa(e_theta_re,'double') );
assertTrue( isa(e_theta_im,'double') );
assertTrue( isa(e_phi_re,'double') );
assertTrue( isa(e_phi_im,'double') );
assertTrue( isa(azimuth_grid,'double') );
assertTrue( isa(elevation_grid,'double') );
assertTrue( isa(element_pos,'double') );
assertTrue( isa(coupling_re,'double') );
assertTrue( isa(coupling_im,'double') );
assertTrue( isa(center_frequency,'double') );
assertTrue( isa(layout,'uint32') );

assertElementsAlmostEqual( 20*log10(e_theta_re), reshape(1:15,3,5)', 'absolute', 1e-14 );
assertElementsAlmostEqual( e_theta_im, zeros(5,3), 'absolute', 1e-14 );
assertElementsAlmostEqual( e_phi_re, zeros(5,3), 'absolute', 1e-14 );
assertElementsAlmostEqual( e_phi_im, zeros(5,3), 'absolute', 1e-14 );
assertElementsAlmostEqual( azimuth_grid, [-pi,0,pi/2], 'absolute', 1e-13 );
assertElementsAlmostEqual( elevation_grid, [-pi/2,-pi/4,0,pi/4,pi/2], 'absolute', 1e-13 );
assertElementsAlmostEqual( element_pos, [0;0;0], 'absolute', 1e-13 );
assertElementsAlmostEqual( coupling_re, 1, 'absolute', 1e-13 );
assertElementsAlmostEqual( coupling_im, 0, 'absolute', 1e-13 );
assertElementsAlmostEqual( center_frequency, 299792448, 'absolute', 1e-13 );
assertTrue( strcmp(name,'bla') );
assertEqual( layout, uint32(1) );

%% More complex test
f = fopen( 'test.qdant','w' );
fprintf(f,'%s\n','<?xml version="1.0" encoding="UTF-8"?><qdant xmlns:xx="test">');
fprintf(f,'%s\n','<xx:layout>1,1 1,1 1,1</xx:layout>');
fprintf(f,'%s\n','<xx:arrayant id="1">');
fprintf(f,'%s\n','<xx:AzimuthGrid>-90 -45 0 45 90</xx:AzimuthGrid>');
fprintf(f,'%s\n','<xx:ElevationGrid>-90 0 90</xx:ElevationGrid>');
fprintf(f,'%s\n',['<xx:EphiMag>',num2str(zeros(1,15)),'</xx:EphiMag>']);
fprintf(f,'%s\n',['<xx:EphiPhase>',num2str(ones(1,15)*90),'</xx:EphiPhase>']);
fprintf(f,'%s\n',['<xx:EthetaMag>',num2str(ones(1,15)*3),'</xx:EthetaMag>']);
fprintf(f,'%s\n',['<xx:EthetaPhase>',num2str(ones(1,15)*-90),'</xx:EthetaPhase>']);
fprintf(f,'%s\n','<xx:ElementPosition>1,2,3</xx:ElementPosition>');
fprintf(f,'%s\n','<xx:CouplingAbs>1</xx:CouplingAbs>');
fprintf(f,'%s\n','<xx:CouplingPhase>45</xx:CouplingPhase>');
fprintf(f,'%s\n','<xx:CenterFrequency>3e9</xx:CenterFrequency>');
fprintf(f,'%s\n','</xx:arrayant></qdant>');
fclose(f);

[e_theta_re, e_theta_im, e_phi_re, e_phi_im, azimuth_grid, elevation_grid, element_pos, ...
    coupling_re, coupling_im, center_frequency, name, layout] = quadriga_lib.arrayant_qdant_read('test.qdant',1,1);

assertTrue( isa(e_theta_re,'single') );
assertTrue( isa(e_theta_im,'single') );
assertTrue( isa(e_phi_re,'single') );
assertTrue( isa(e_phi_im,'single') );
assertTrue( isa(azimuth_grid,'single') );
assertTrue( isa(elevation_grid,'single') );
assertTrue( isa(element_pos,'single') );
assertTrue( isa(coupling_re,'single') );
assertTrue( isa(coupling_im,'single') );
assertTrue( isa(center_frequency,'single') );

assertElementsAlmostEqual( azimuth_grid, [-pi/2,-pi/4,0,pi/4,pi/2], 'absolute', 1e-7 );
assertElementsAlmostEqual( elevation_grid, [-pi/2,0,pi/2], 'absolute', 1e-7 );
assertElementsAlmostEqual( e_theta_re, zeros(3,5), 'absolute', 1e-7 );
assertElementsAlmostEqual( e_theta_im, -sqrt(10^(0.3)*ones(3,5)), 'absolute', 1e-6 );
assertElementsAlmostEqual( e_phi_re, zeros(3,5), 'absolute', 1e-7 );
assertElementsAlmostEqual( e_phi_im, ones(3,5), 'absolute', 1e-7 );
assertElementsAlmostEqual( element_pos, [1;2;3], 'absolute', 1e-7 );
assertElementsAlmostEqual( coupling_re, 1/sqrt(2), 'absolute', 1e-7 );
assertElementsAlmostEqual( coupling_im, 1/sqrt(2), 'absolute', 1e-7 );
assertElementsAlmostEqual( center_frequency, 3e9, 'absolute', 1e-13 );
assertTrue( strcmp(name,'unknown') );
assertEqual( layout, ones(2,3,'uint32') );

%% Two array antennas with uncommon formats
f = fopen( 'test.qdant','w' );
fprintf(f,'<qdant><arrayant id="1">\n');
fprintf(f,'<ElevationGrid> -45 45</ElevationGrid>\n');
fprintf(f,'<AzimuthGrid>-90 0 90</AzimuthGrid>\n');
fprintf(f,'<EthetaMag>\n\t1 1 1\n\t2 2 3\n</EthetaMag>\n');
fprintf(f,'</arrayant><arrayant id="3">\n');
fprintf(f,'<ElevationGrid> -45 45</ElevationGrid>');
fprintf(f,'<NoElements>2</NoElements>');
fprintf(f,'<AzimuthGrid>-90 0 90</AzimuthGrid>\n');
fprintf(f,'<EphiMag el="2">\n\t1 2 3\n\t-1 -2 -3\n</EphiMag>\n');
fprintf(f,'<EphiPhase el="2">90 90 90 -90 -90 -90</EphiPhase>\n');
fprintf(f,'<CouplingAbs>1 2</CouplingAbs>\n');
fprintf(f,'<CouplingPhase>45 -90</CouplingPhase>\n');
fprintf(f,'<name>xxx</name>\n');
fprintf(f,'</arrayant></qdant>\n');
fclose(f);

[e_theta_re, e_theta_im, e_phi_re, e_phi_im, azimuth_grid, elevation_grid, element_pos, ...
    coupling_re, coupling_im, center_frequency, name, layout] = quadriga_lib.arrayant_qdant_read('test.qdant',1);

assertEqual( layout, uint32([1 3]) );

assertElementsAlmostEqual( e_theta_re, sqrt([1.26 1.26 1.26 ; 1.58 1.58 2]), 'absolute', 1e-2 );

[e_theta_re, e_theta_im, e_phi_re, e_phi_im, azimuth_grid, elevation_grid, element_pos, ...
    coupling_re, coupling_im, center_frequency, name] = quadriga_lib.arrayant_qdant_read('test.qdant',3);

assertElementsAlmostEqual( e_theta_re, zeros(2,3,2), 'absolute', 1e-13 );
assertElementsAlmostEqual( e_phi_re, zeros(2,3,2), 'absolute', 1e-13 );
assertElementsAlmostEqual( e_phi_im(:,:,2), [1;-1].*sqrt(10.^([1 2 3 ; -1 -2 -3]/10)), 'absolute', 1e-13 );
assertElementsAlmostEqual( coupling_re, [1/sqrt(2);0], 'absolute', 1e-13 );
assertElementsAlmostEqual( coupling_im, [1/sqrt(2);-2], 'absolute', 1e-13 );
assertTrue( strcmp(name,'xxx') );

%% Check error parsing
try
    [~,~,~,~,~,~,~,~,~,~,~] = quadriga_lib.arrayant_qdant_read('test.qdant',3);
    error('moxunit:exceptionNotRaised', 'Expected an error!');
end

% AzimuthGrid missing
f = fopen( 'test.qdant','w' );
fprintf(f,'<qdant><arrayant>\n');
fprintf(f,'<ElevationGrid>-90 -45 0 45 90</ElevationGrid>\n');
fprintf(f,'</arrayant></qdant>');
fclose(f);

try
    [~,~,~,~,~,~,~,~,~,~,~] = quadriga_lib.arrayant_qdant_read('test.qdant');
    error('moxunit:exceptionNotRaised', 'Expected an error!');
end

% AzimuthGrid tag not closed correctly
f = fopen( 'test.qdant','w' );
fprintf(f,'<qdant><arrayant>\n');
fprintf(f,'<ElevationGrid>-90 -45 0 45 90</ElevationGrid>\n');
fprintf(f,'<AzimuthGrid>-90 0 90</AzimuthGridXXX>\n');
fprintf(f,'</arrayant></qdant>');
fclose(f);

try
    [~,~,~,~,~,~,~,~,~,~,~] = quadriga_lib.arrayant_qdant_read('test.qdant');
    error('moxunit:exceptionNotRaised', 'Expected an error!');
end

% Wrong number of entries in EthetaMag
f = fopen( 'test.qdant','w' );
fprintf(f,'<qdant><arrayant>\n');
fprintf(f,'<ElevationGrid>-90 -45 0 45 90</ElevationGrid>\n');
fprintf(f,'<AzimuthGrid>-90 0 90</AzimuthGrid>\n');
fprintf(f,'%s\n',['<EthetaMag>',num2str(1:16),'</EthetaMag>']);
fprintf(f,'</arrayant></qdant>');
fclose(f);

try
    [~,~,~,~,~,~,~,~,~,~,~] = quadriga_lib.arrayant_qdant_read('test.qdant');
    error('moxunit:exceptionNotRaised', 'Expected an error!');
end

% Wrong number of entries in CouplingAbs
f = fopen( 'test.qdant','w' );
fprintf(f,'<qdant><arrayant><NoElements>2</NoElements>\n');
fprintf(f,'<ElevationGrid>-90 -45 0 45 90</ElevationGrid>\n');
fprintf(f,'<AzimuthGrid>-90 0 90</AzimuthGrid>\n');
fprintf(f,'%s\n',['<EthetaMag>',num2str(1:15),'</EthetaMag>']);
fprintf(f,'<CouplingAbs>1 2 3</CouplingAbs>\n');
fprintf(f,'</arrayant></qdant>');
fclose(f);

try
    [~,~,~,~,~,~,~,~,~,~,~] = quadriga_lib.arrayant_qdant_read('test.qdant');
    error('moxunit:exceptionNotRaised', 'Expected an error!');
end

% CouplingPhase with multiple ports without CouplingAbs
f = fopen( 'test.qdant','w' );
fprintf(f,'<qdant><arrayant>\n');
fprintf(f,'<ElevationGrid>-90 -45 0 45 90</ElevationGrid>\n');
fprintf(f,'<AzimuthGrid>-90 0 90</AzimuthGrid>\n');
fprintf(f,'<CouplingPhase>90 90</CouplingPhase>\n');
fprintf(f,'</arrayant></qdant>');
fclose(f);

try
    [~,~,~,~,~,~,~,~,~,~,~] = quadriga_lib.arrayant_qdant_read('test.qdant');
    error('moxunit:exceptionNotRaised', 'Expected an error!');
end

% Not a XML File
f = fopen( 'test.qdant','w' );
fprintf(f,'bla bla bla\n');
fclose(f);

try
    [~,~,~,~,~,~,~,~,~,~,~] = quadriga_lib.arrayant_qdant_read('test.qdant');
    error('moxunit:exceptionNotRaised', 'Expected an error!');
end

% Not a QDANT file (KML instead)
f = fopen( 'test.qdant','w' );
fprintf(f,'<?xml version="1.0" encoding="UTF-8"?>\n');
fprintf(f,'<kml xmlns="http://www.opengis.net/kml/2.2">\n');
fprintf(f,'  <Placemark>\n');
fprintf(f,'    <name>A simple placemark on the ground</name>\n');
fprintf(f,'    <Point>\n');
fprintf(f,'		<coordinates>8.542952335953721,47.36685263064198,0</coordinates>\n');
fprintf(f,'    </Point>\n');
fprintf(f,'  </Placemark>\n');
fprintf(f,'</kml>\n');
fclose(f);

try
    [~,~,~,~,~,~,~,~,~,~,~] = quadriga_lib.arrayant_qdant_read('test.qdant');
    error('moxunit:exceptionNotRaised', 'Expected an error!');
end

delete('test.qdant');


